experimental/Schemes/Tjurina.jl

```diff
@@ -0,0 +1,770 @@
+  return tjurina_algebra(gen(modulus(OO(X)),1))
+end
+
+
+
+@doc raw"""
+    tjurina_algebra(X::HypersurfaceGerm)
+    tjurina_algebra(X::HypersurfaceGerm, k::Integer = 0)
+
+Return the `k`-th local Tjurina algebra of `(X,p)` at `p`. 
+By default computes the local Tjurina algebra (`k=0`) at `p`.
+Higher Tjurina algebras are of interest in positive characteristic.
+# Examples
+```jldoctest
+julia> R,(x,y) = QQ["x", "y"];
+
+julia> f = x^3-y^2;
+
+julia> X = HypersurfaceGerm(AffineScheme(quo(R, ideal(R, f))[1]), [0, 0]);
+
+julia> tjurina_algebra(X)
+Localization
+  of quotient
+    of multivariate polynomial ring in 2 variables x, y
+      over rational field
+    by ideal (x^3 - y^2, 3*x^2, -2*y)
+  at complement of maximal ideal of point (0, 0)
+```
+"""
+function tjurina_algebra(X::HypersurfaceGerm, k::Integer = 0)  
+  k >= 0 || error("Integer must be non-negative.")
+  R = localized_ring(OO(X))
+  ## tjurina algebra independent of choice of representative
+  ## hence choose a polynomial representative for easier computation
+  f_poly = numerator(defining_ideal(X)[1])
+  m = ideal(R, gens(R)-point(X))
+  I = ideal(R, f_poly) + m^k*ideal(R, R.([derivative(f_poly, i) for i=1:nvars(base_ring(R))]))
+  return quo(R,I)[1]
+end
+
+
+
+@doc raw"""
+    tjurina_algebra(f::MPolyLocRingElem{<:Any, <:Any, <:Any, <:Any, <:MPolyComplementOfKPointIdeal})
+    tjurina_algebra(f::MPolyLocRingElem{<:Any, <:Any, <:Any, <:Any, <:MPolyComplementOfKPointIdeal}, k::Integer = 0)
+
+Return the `k`-th local Tjurina algebra of `(V(f),p)` at `p`. 
+By default computes the local Tjurina algebra (`k=0`) at `p`.
+Higher Tjurina algebras are of interest in positive characteristic.
+# Examples
+```jldoctest
+julia> R,(x,y) = GF(2)["x", "y"];
+
+julia> L,_  = localization(R, complement_of_point_ideal(R, [0, 0]));
+
+julia> tjurina_algebra(L(x^3-y^2))
+Localization
+  of quotient
+    of multivariate polynomial ring in 2 variables x, y
+      over prime field of characteristic 2
+    by ideal (x^3 + y^2, x^2, 0)
+  at complement of maximal ideal of point (0, 0)
+
+julia> tjurina_algebra(L(x^3-y^2), 1)
+Localization
+  of quotient
+    of multivariate polynomial ring in 2 variables x, y
+      over prime field of characteristic 2
+    by ideal (x^3 + y^2, x^3, 0, x^2*y, 0)
+  at complement of maximal ideal of point (0, 0)
+```
+"""
+function tjurina_algebra(f::MPolyLocRingElem{<:Any, <:Any, <:Any, <:Any, <:MPolyComplementOfKPointIdeal}, k::Integer = 0)
+  X = HypersurfaceGerm(quo(parent(f), ideal(parent(f), f))[1])
+  return tjurina_algebra(X, k)
+end
+
+
+
+
+
+###############################################################################
+
+#####                           Tjurina number                            #####
+
+###############################################################################
+
+@doc raw"""
+    tjurina_number(f::MPolyRingElem)
+
+Return the global Tjurina number of a polynomial `f`. Return -1 if infinite.
```

Comment:
I really don't like this convention of returning `-1` to indicate infinity. That's gone cause all kinds of bugs down the line.

Alternatives we use elsewhere include:
- return an actual infinite (via `PosInf()`); 
- raise a (dedicated) exception indicating infinite order
- add a separate test function like `has_finite_tjurina_number`

Reply:
An exception would raise all kinds of complications down the road, I would rather use infinity, as this is the mathematically correct statement.
has_finite_tjurina_number has no added value here, because this is synonymous with is_finitely_determined by the theorem of Mather-Yau and the test for this is available.

Reply:
I totally agree with using PosInf() instead of an exception. But I think the problem lies in vector_space_dimension(A::MPolyQuo) throwing an exception instead of returning infinity and therefore should be fixed at that level. 
Should I open a feature request or fix that issue in this PR?

Reply:
I wouldn't do it in this pull request.
Tinkering with the output of `vector_space_dimension` would be a breaking change to my understanding and we are not allowed to do these any more since we released 1.0. It would break the code of people who did a similar try-catch work around to catch an infinite dimensional ring if there is no error.
But that's just my two cents.

Reply:
I would prefer that you open a feature request issue for vectore_space_dimension and use positive infinity here in the meantime, to make sure that everything building on top of it does not break in the future, when the feature is changed.